pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,243 @@
+func (w *Watcher) traversePluginDir() error {
+	files, err := w.fs.ReadDir(w.path)
+	if err != nil {
+		return fmt.Errorf("error reading the plugin directory: %v", err)
+	}
+	for _, f := range files {
+		// Currently only supports flat fs namespace under the plugin directory.
+		// TODO: adds support for hierarchical fs namespace.
+		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
+			w.registerPlugin(path.Join(w.path, f.Name()))
+		}
+	}
+	return nil
+}
+
+func (w *Watcher) init() error {
+	if err := w.createPluginDir(); err != nil {
+		return err
+	}
+	return w.traversePluginDir()
+}
+
+func (w *Watcher) registerPlugin(socketPath string) error {
+	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
+	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		return fmt.Errorf("dial failed at socket %s, err: %v", socketPath, err)
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	infoResp, err := client.GetInfo(ctx, &registerapi.InfoRequest{})
+	if err != nil {
+		glog.Errorf("Failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	resp := w.invokeRegistrationCallbackAtHandler(infoResp, socketPath)
+	glog.V(2).Infof("call NotifyRegistrationStatus, resp: %v", resp)
+	if _, err := client.NotifyRegistrationStatus(ctx, resp); err != nil {
+		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	if !resp.PluginRegistered {
+		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", infoResp.Type, infoResp.Name, socketPath, resp.Error)
+		return fmt.Errorf("%s", resp.Error)
+	}
+	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", infoResp.Type, infoResp.Name, socketPath)
+	return nil
+}
+
+func (w *Watcher) invokeRegistrationCallbackAtHandler(infoResp *registerapi.PluginInfo, socketPath string) *registerapi.RegistrationStatus {
+	var handlerCbkFn RegisterCallbackFn
+	var ok bool
+	handlerCbkFn, ok = w.handlers[infoResp.Type]
+	if !ok {
+		return &registerapi.RegistrationStatus{
+			PluginRegistered: false,
+			Error:            fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", infoResp.Type, socketPath),
+		}
+	}
+
+	var versions []string
+	for _, version := range infoResp.SupportedVersions {
+		versions = append(versions, version)
+	}
+	// calls handler callback to verify registration request
+	if err := handlerCbkFn(infoResp.Name, versions, socketPath); err != nil {
+		return &registerapi.RegistrationStatus{
+			PluginRegistered: false,
+			Error:            fmt.Sprintf("Plugin registration failed with err: %v", err),
+		}
+	}
+
+	return &registerapi.RegistrationStatus{PluginRegistered: true}
+}
+
+// Start watches for the creation of plugin sockets at the path
```

Comment:
this is not about your patch, but about the design of watching fsnotify events, 
this approach is relying on an assumption that: fsnofity will be reliably notify pluginwatcher.
but i am just playing evil here, this assumption maybe not 100% true by looking at number of open issue:
https://github.com/fsnotify/fsnotify/issues?page=3&q=is%3Aissue+is%3Aopen and my local test on Mac also show issue with fsnotify.

the other possible workaround is to traverse folder with interval, how do you think?

Reply:
I'm not sure the number of issue is a reliable metric as many issue seem to be questions , feature requests or refer to an old version :)

Additionally I believe fsnotify is production ready given [the number of projects already using it](https://github.com/search?p=1&q=%22github.com%2Ffsnotify%2Ffsnotify%22&type=Code) (including docker and k8s)

Reply:
Given the number of plugins we expect is small, traversal might not be a bad choice. However inotify does provide better latency though and avoids unnecessary traversals. @vikaschoudhary16 it would be worth examining the open issues against inotify prior to taking a hard dependency on it.

Reply:
thanks @figo for pointing this out.
@vishh I took a look around this. It seems fsnotify is not being maintained actively. And reading through github issues, looks like people have been maintaining there forks as well for the some of their fixes which could not get merged. Here are the relevant links:
https://github.com/fsnotify/fsnotify/issues/245#issuecomment-388498752

On the other hand, what @RenaudWasTaken said is also true that so many projects are using it. And therefore i think there are chances that this project will likely be able to find a adopter soon. I am slightly in favor of keeping fsnotify, at the same time if majority thinks traversal might be a better option, open to make changes for traversal as well.

@jiayingz @saad-ali @thockin what do you guys think?

Reply:
I'd prefer not taking on a dependency that is not being maintained actively. We should either have a plan to contribute back to that project or fork it in k8s organization.

Reply:
I would prefer to staying with fsnotify for now. Mostly we just need a go pkg for inotify syscall. The fsnotify pkg is already included in k8s so we are not adding a new dependency, and so far we haven't heard issues in the past for the included version. It will be a much easier change to switch to a more stable fsnotify go pkg in the future than building on top of a different model where the latency to detect a new plugin would be affected by how long and how often directory traverse happens.